File: C++/Searching and Sorting/12_stickler_thief.cpp
/*
Idea:
there are 2 options in a single element
-choose or not choose, if element is chosen, move the pointer 2 steps ahead, else only 1.
-using vector to store current max sum till that element
*/

#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;

class Solution
{
public:
    int FindMaxSum(int arr[], int n)
    {
        if (n == 1)return arr[0]; // if array has only 1 element

        vector<int>dp(n, 0);
        dp[0] = arr[0];
        dp[1] = max(arr[0], arr[1]); // setting up max with first 2 elements.

        for (int i = 2; i < n; i++)
            dp[i] = max(dp[i - 1], dp[i - 2] + arr[i]);
        // 2 cases, if element is added or not.

        return dp[n - 1];
    }
};


int main()
{
    //taking total testcases
    int t;
    cin >> t;
    while (t--)
    {
        //taking number of houses
        int n;
        cin >> n;
        int a[n];

        //inserting money of each house in the array
        for (int i = 0;i < n;++i)
            cin >> a[i];
        Solution ob;
        //calling function FindMaxSum()
        cout << ob.FindMaxSum(a, n) << endl;
    }
    return 0;
}

File: Python/Searching and Sorting/12_stickler_thief.py
"""
Idea:
there are 2 options in a single element
-choose or not choose, if element is chosen, move the pointer 2 steps ahead, else only 1.
-using list to store current max sum till that element
"""


class Solution:

    def FindMaxSum(self, a, n):
        if n == 1:  # if array has only 1 element
            return a[0]

        dp = [0] * n
        dp[0] = a[0]
        dp[1] = max(a[0], a[1])  # setting up max with first 2 elements.
        for i in range(2, n):
            dp[i] = max(dp[i - 1], dp[i - 2] + a[i])
            # 2 cases, if element is added or not.
        return dp[n - 1]


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        a = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.FindMaxSum(a, n))
